surface/include/pcl/surface/impl/mls.hpp

```diff
@@ -673,45 +534,363 @@ pcl::MovingLeastSquares<PointInT, PointOutT>::performUpsampling (PointCloudOut &
-
-      // Store the id of the original point
-      corresponding_input_indices_->indices.push_back (input_index);
-
-      output.push_back (result_point);
-
-      if (compute_normals_)
-        normals_->push_back (result_normal);
+      double u, v, w;
+      getMLSCoordinates (add_point, mls_results_[input_index], u, v, w);
+
+      MLSProjectionResults proj;
+      if (polynomial_fit_ && mls_results_[input_index].num_neighbors >= 5 * nr_coeff_)
+      {
+        if (projection_method_ == SIMPLE)
+          proj = projectPointSimpleToPolynomialSurface (u, v, mls_results_[input_index]);
+        else if (projection_method_ == ORTHOGONAL)
+          proj = projectPointOrthogonalToPolynomialSurface (u, v, w, mls_results_[input_index]);
+        else
+          proj = projectPointToMLSPlane (u, v, mls_results_[input_index]);
+      }
+      else
+      {
+        proj = projectPointToMLSPlane (u, v, mls_results_[input_index]);
+      }
+
+      addProjectedPointNormal(input_index, proj.point, proj.normal, mls_results_[input_index].curvature, output, *normals_, *corresponding_input_indices_);
     }
   }
 }
 
 //////////////////////////////////////////////////////////////////////////////////////////////
-pcl::MLSResult::MLSResult (const Eigen::Vector3d &a_mean,
-                             const Eigen::Vector3d &a_plane_normal,
-                             const Eigen::Vector3d &a_u,
-                             const Eigen::Vector3d &a_v,
-                             const Eigen::VectorXd a_c_vec,
-                             const int a_num_neighbors,
-                             const float &a_curvature) :
-  mean (a_mean), plane_normal (a_plane_normal), u_axis (a_u), v_axis (a_v), c_vec (a_c_vec), num_neighbors (a_num_neighbors),
-  curvature (a_curvature), valid (true)
+pcl::MLSResult::MLSResult (const Eigen::Vector3d &a_query_point,
+                           const Eigen::Vector3d &a_mean,
+                           const Eigen::Vector3d &a_plane_normal,
+                           const Eigen::Vector3d &a_u,
+                           const Eigen::Vector3d &a_v,
+                           const Eigen::VectorXd &a_c_vec,
+                           const int a_num_neighbors,
+                           const float &a_curvature,
+                           const int a_order,
+                           const bool a_polynomial_fit):
+  query_point (a_query_point), mean (a_mean), plane_normal (a_plane_normal), u_axis (a_u), v_axis (a_v), c_vec (a_c_vec), num_neighbors (a_num_neighbors),
+  curvature (a_curvature), order (a_order), polynomial_fit(a_polynomial_fit), valid (true)
+{
+}
+
+void pcl::getMLSCoordinates (const Eigen::Vector3d &pt, const pcl::MLSResult &mls_result, double &u, double &v, double &w)
+{
+  u = (pt - mls_result.mean).dot (mls_result.u_axis);
+  v = (pt - mls_result.mean).dot (mls_result.v_axis);
+  w = (pt - mls_result.mean).dot (mls_result.plane_normal);
```

Comment:
Cache the subtraction first and then apply the dot products on it. Similar below.